### PR TITLE
Dijkstra's single source shortest path algorithm

### PR DESCRIPTION
**Pull request by agranya99**
A popular algorithm that returns shortest path to all vertices from the source vertex.

### CODE REVIEW

**Comment by danghai** [comment]
Good! Could you provide the `unittest` for this algorithm?

**Comment by agranya99** [comment]
Sure, will do it.

File: graph/dij.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import sys
+ 
+class Graph():
+
+    def __init__(self, vertices):
+        self.V = vertices
+        self.graph = [[0 for column in range(vertices)] 
+                      for row in range(vertices)]
+ 
+    def printSolution(self, dist):
```
Comment (goswami-rahul) [inline graph/dij.py]: we don't need this printSolution function here.

File: graph/dij.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        print("Vertex \t\t Min Distance from Source")
+        for node in range(self.V):
+            print(node,"   \t\t",dist[node])
+ 
+
+    def minDistance(self, dist, MinSet):
```
Comment (goswami-rahul) [inline graph/dij.py]: 'camelCase' is not preferred for function and variable names. Use 'snake_case' like `min_distance`.

File: graph/dij.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                min_index = v
+ 
+        return min_index
+ 
+
+    def dijkstra(self, src):
+ 
+        dist = [sys.maxsize] * self.V
+        dist[src] = 0
+        MinSet = [False] * self.V
```
Comment (goswami-rahul) [inline graph/dij.py]: rename `MinSet` too.

File: graph/dij.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+#Dijkstra's single source shortest path algorithm
+
+import sys
+ 
+class Graph():
+
+    def __init__(self, vertices):
+        self.V = vertices
```
Comment (goswami-rahul) [inline graph/dij.py]: `self.V` could be renamed to more specific `self.vertices`.

File: graph/dij.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            if dist[v] < min and MinSet[v] == False:
+                min = dist[v]
+                min_index = v
+ 
+        return min_index
+ 
+
+    def dijkstra(self, src):
+ 
+        dist = [sys.maxsize] * self.V
```
Comment (goswami-rahul) [inline graph/dij.py]: no need for `sys.maxsize`. Use `float("inf")`. And remove import sys.

File: graph/dij.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ 
+            # Update dist value of the adjacent vertices 
+            for v in range(self.V):
+                if self.graph[u][v] > 0 and MinSet[v] == False and dist[v] > dist[u] + self.graph[u][v]:
+                        dist[v] = dist[u] + self.graph[u][v]
+ 
+        self.printSolution(dist)
+ 
+#MAIN
+"""vert=input("Enter the number of vertices: ")
```
Comment (goswami-rahul) [inline graph/dij.py]: remove all part below this. not required in this file.

**Comment by danghai** [comment]
@goswami-rahul shoud we add the reference about the source?

**Comment by agranya99** [comment]
@goswami-rahul Thanks for the suggestions. Have made the suggested changes.

File: stack/orderedStack.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+         return self.items.pop()
+         
+     def peek(self):
+         return self.items[len(self.items)-1]
+         
+     def size(self):
+         return len(self.items)
+         
+     def printStack(self):
+	     sz=self.size()
```
Comment (danghai) [inline stack/orderedStack.py]: It should be `sz = self.size()`

File: stack/orderedStack.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+             raise IndexError("Stack is empty")
+         return self.items.pop()
+         
+     def peek(self):
+         return self.items[len(self.items)-1]
+         
+     def size(self):
+         return len(self.items)
+         
+     def printStack(self):
```
Comment (danghai) [inline stack/orderedStack.py]: we may not need the method `printStack`, could you remove it?

File: stack/orderedStack.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+#The stack remains always ordered such that the highest value is at the top and the lowest at the bottom
+
+class Stack:
```
Comment (danghai) [inline stack/orderedStack.py]: Could you rename `Stack` with different name `Stack` for clear?

**Comment by danghai** [comment]
Nice addition of Stack! . But some changes are required. Thanks

File: stack/orderedStack.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+         self.items = []
+         
+     def isEmpty(self):
+         return self.items == []
+         
+     def pushT(self, item):
+         self.items.append(item)
+	     
+     def push(self, item): #push method to maintain order when pushing new elements
+         temp=Stack()
```
Comment (danghai) [inline stack/orderedStack.py]: It should be `temp = Stack()`

**Comment by danghai** [comment]
@agranya99 I update something in your `order_stack`.

File: graph/dijkstra.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+class Graph():
+
+    def __init__(self, vertices):
+        self.vertices = vertices
+        self.graph = [[0 for column in range(vertices)]
+                      for row in range(vertices)]
+
+    def min_distance(self, dist, min_dist_set):
+        min = float("inf")
```
Comment (goswami-rahul) [inline graph/dijkstra.py]: Also need to rename this `min` -> `min_dist`. `min` is a buit-in function name.

File: stack/ordered_stack.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
+#The stack remains always ordered such that the highest value is at the top and the lowest at the bottom
+
+class OrderedStack:
+     def __init__(self):
+         self.items = []
+
+     def isEmpty(self):
```
Comment (goswami-rahul) [inline stack/ordered_stack.py]: rename to `is_empty()`.

File: stack/ordered_stack.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+         return self.items == []
+
+     def pushT(self, item):
```
Comment (goswami-rahul) [inline stack/ordered_stack.py]: rename to `push_t`.

File: stack/ordered_stack.py [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+     def push(self, item): #push method to maintain order when pushing new elements
+         temp = OrderedStack()
+         if self.isEmpty() or item > self.peek():
+             self.pushT(item)
+             return self.items
+         else:
+             while (item < self.peek()) and (not self.isEmpty()):
+                 temp.pushT(self.pop())
+             self.pushT(item)
+             while not(temp.isEmpty()):
```
Comment (goswami-rahul) [inline stack/ordered_stack.py]: no need for outside brackets.

File: stack/ordered_stack.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+     def pushT(self, item):
+         self.items.append(item)
+
+     def push(self, item): #push method to maintain order when pushing new elements
+         temp = OrderedStack()
+         if self.isEmpty() or item > self.peek():
+             self.pushT(item)
+             return self.items
+         else:
+             while (item < self.peek()) and (not self.isEmpty()):
```
Comment (goswami-rahul) [inline stack/ordered_stack.py]: no need for redundant brackets.

File: stack/ordered_stack.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+                 temp.pushT(self.pop())
+             self.pushT(item)
+             while not(temp.isEmpty()):
+                 self.pushT(temp.pop())
+             return self.items
```
Comment (goswami-rahul) [inline stack/ordered_stack.py]: `self.push(item)` must return `None` (not `self.tems`)

File: tests/test_stack.py [hunk 17]
```diff
@@ -129,5 +130,19 @@ def test_LinkedListStack(self):
...
 
+class TestOrderedStack(unittest.TestCase):
+    def test_OrderedStack(self):
+        stack = OrderedStack()
+        self.assertTrue(stack.isEmpty())
+        stack.push(1)
+        stack.push(4)
+        stack.push(3)
+        "bottom - > 1 3 4 6 "
+        self.assertEqual([1, 3, 4, 6], stack.push(6))
```
Comment (goswami-rahul) [inline tests/test_stack.py]: @danghai Shouldn't `stack.push(item)` return `None`. We can't test for it to be equal to the list.

**Comment by goswami-rahul** [comment]
@agranya99 The tests are failed because of some indentation error. Please use 4 spaces everywhere, and don't mix them with tabs (see your IDE settings for this).

File: tests/test_stack.py (hunk 17 above)
Comment (agranya99) [reply tests/test_stack.py]: Yes, the unittest was made by one of the collaborators for the previous code. I have removed return statement. Will modify the unittest

**Pull request merged**